HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+For a trivial example like this, the cost and efficiency deltas may not seem decisive (unless the project was
+running a very large number of these puzzles). But the idea of releasing contracts from the burden of duplicating
+native protocol logic is deceptively powerful. The cost and efficiency savings for a complex dApp could be enormous,
+unlocking entire new classes of applications that would be impractical to build on Hedera today.
+
+#### Receiver signature waiver for HTS assets without custom fees
+
+In this example we have our own account `0.0.Y` with `receiver_sig_required=true`, and want to carve out an exception
+for exactly HTS token credits to our account with no assessed custom fees. We install a pure EVM hook at index `2`
+whose referenced contract is as follows,
+
+```solidity
+import "./IHederaTokenService.sol";
+import "./IHieroAccountAllowanceEvmHook.sol";
+
+contract CreditSansCustomFeesTokenAllowance is IHieroAccountAllowanceEvmHook {
+    /// Allows the proposed transfers only if,
+    ///   (1) The only transfers are direct HTS asset transfers
+    ///   (2) The installer is not debited
+    ///   (3) The installer is credited
+    function allow(
+        address installer,
+        IHieroTransferAllowance.ProposedTransfers memory proposedTransfers,
+        bytes memory args
+    ) external override view returns (bool) {
+        require(address(this) == 0x16d, "Contract can only be called as a hook");
+        if (proposedTransfers.direct.hbar.transfers.length > 0
+                || proposedTransfers.customFee.hbar.transfers.length > 0
+                || proposedTransfers.customFee.tokens.length > 0) {
+            return false;
+        }
+        bool installerCredited = false;
+        for (uint256 i = 0; i < proposedTransfers.tokens.length; i++) {
+            IHederaTokenService.AccountAmount[] memory transfers = proposedTransfers.tokens[i].transfers;
+            for (uint256 j = 0; j < transfers.length; j++) {
+                if (transfers[j].accountID == installer) {
+                    if (transfers[j].amount < 0) {
+                        return false;
+                    } else if (transfers[j].amount > 0) {
+                        installerCredited = true;
+                    }
+                }
+            }
+            IHederaTokenService.NftTransfer[] memory nftTransfers = proposedTransfers.tokens[i].nftTransfers;
+            for (uint256 j = 0; j < nftTransfers.length; j++) {
+                if (nftTransfers[j].senderAccountID == installer) {
+                    return false;
+                } else if (nftTransfers[j].receiverAccountID == installer) {
+                    installerCredited = true;
+                }
+            }
+        }
+        return installerCredited;
+    }
+}
+```
+
+## Backwards Compatibility
+
+This HIP adds a net new feature to the protocol. Any account that does not install a hook will see
+identical behavior in all circumstances.
+
+## Security Implications
```

Comment:
In my opinion, in this paragraph it is necessary to clearly state and document all privileges that the hook or hook system contract will have during execution, how external contract will see them, etc.

Reply:
Covered in more detail in [this](https://github.com/tinker-michaelj/hiero-improvement-proposals/blob/main/HIP/hip-1195.md#the-evm-environment-for-hooks) new section.